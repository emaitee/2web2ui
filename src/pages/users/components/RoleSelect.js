import React from 'react';
import { Select } from '@sparkpost/matchbox';
import _ from 'lodash';

const DEFAULT_OPTIONS = [
  { label: 'Admin', value: 'admin' },
  { label: 'Reporting', value: 'reporting' }
];

/**
 * Dropdown to select a users access/role
 *
 * @param {boolean} disabled
 * @param {string} name
 * @param {Function} onChange - event handler passes new
 * @param {null, string} value
 * @param {Object} input - redux-form <Field /> props
 */
export default function RoleSelect({
  disabled = false,
  name = null,
  onChange = _.noop,
  value = null,
  input = {},
  allowSuperUser = false,
  selectOptions = [ ...DEFAULT_OPTIONS ],
  ...rest
}) {

  if (allowSuperUser) {
    selectOptions.push({ label: 'Super User', value: 'superuser' });
  }

  // Only return text instead of select box
  // TODO: won't need the span tags with v16
  if (disabled) {
    const label = selectOptions.find((option) => option.value === value).label;
    return <span>{label}</span>;
  }

  const handleChange = ({ target }) => onChange(target);

  return (
    <Select
      name={name}
      onChange={handleChange}
      options={selectOptions}
      value={value}
      {...input}
      {...rest}
    />
  );
}
