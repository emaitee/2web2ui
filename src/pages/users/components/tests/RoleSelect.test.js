import { shallow } from 'enzyme';
import React from 'react';
import { Select } from '@sparkpost/matchbox';
import RoleSelect from '../RoleSelect';

describe('Component: RoleSelect', () => {

  let props;
  let wrapper;

  beforeEach(() => {
    props = { value: 'admin' };
    wrapper = shallow(<RoleSelect {...props} />);
  });

  it('should render the select', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render when disabled', () => {
    wrapper.setProps({ disabled: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with default select options', () => {
    const options = wrapper.find(Select).props().options;
    expect(options.map((o) => o.value)).toEqual(['admin', 'reporting']);
  });

  it('should render with passed in select options', () => {
    wrapper.setProps({ selectOptions: [
      { label: 'A', value: 'a' },
      { label: 'B', value: 'b' }
    ]});
    const options = wrapper.find(Select).props().options;
    expect(options.map((o) => o.value)).toEqual(['a', 'b']);
  });

  it('should include superuser as a select option if specified in props', () => {
    wrapper.setProps({ allowSuperUser: true });
    const options = wrapper.find(Select).props().options;
    expect(options.map((o) => o.value)).toEqual(['admin', 'reporting', 'superuser']);
  });

  it('should call the onChange handler with the target of the event', () => {
    const target = {};
    const e = { target };
    const onChange = jest.fn();

    wrapper.setProps({ onChange });
    wrapper.find(Select).simulate('change', e);

    expect(onChange).toHaveBeenCalledWith(target);
  });

});
