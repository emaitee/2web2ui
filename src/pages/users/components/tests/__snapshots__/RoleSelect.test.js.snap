// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: RoleSelect should render the select 1`] = `
<Select
  name={null}
  onChange={[Function]}
  options={
    Array [
      Object {
        "label": "Admin",
        "value": "admin",
      },
      Object {
        "label": "Reporting",
        "value": "reporting",
      },
    ]
  }
  placeholderValue=""
  required={false}
  value="admin"
/>
`;

exports[`Component: RoleSelect should render when disabled 1`] = `
<span>
  Admin
</span>
`;
